kratos/sources/table_accessor.cpp

```diff
@@ -0,0 +1,105 @@
+
+namespace Kratos
+{
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+double TableAccessor::GetValue(
+    const Variable<double>& rVariable,
+    const Properties& rProperties,
+    const GeometryType& rGeometry,
+    const Vector& rShapeFunctionVector,
+    const ProcessInfo& rProcessInfo
+    ) const
+{
+    return GetValueFromTable(*mpInputVariable, rVariable, rProperties, rGeometry, rShapeFunctionVector, rProcessInfo);
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+double TableAccessor::GetValueFromTable(
+        const Variable<double> &rIndependentVariable,
+        const Variable<double> &rDependentVariable,
+        const Properties& rProperties,
+        const GeometryType& rGeometry,
+        const Vector& rShapeFunctionVector,
+        const ProcessInfo& rProcessInfo) const
+{
+    double independent_at_gauss = 0.0;
+    if (mInputVariableType == Globals::DataLocation::NodeHistorical) {
+        KRATOS_ERROR_IF_NOT(rGeometry[0].SolutionStepsDataHas(rIndependentVariable)) << "The Variable " << rIndependentVariable.Name() << " is not available at the nodes of the Geometry to retrieve Table values." << std::endl;
+        for (SizeType i = 0; i < rShapeFunctionVector.size(); ++i) {
+            KRATOS_DEBUG_ERROR_IF_NOT(rGeometry[i].SolutionStepsDataHas(rIndependentVariable)) << "The Variable " << rIndependentVariable.Name() << " is not available at the nodes of the Geometry to retrieve Table values." << std::endl;
+            const double nodal_value = rGeometry[i].FastGetSolutionStepValue(rIndependentVariable);
+            independent_at_gauss += nodal_value * rShapeFunctionVector[i];
+        }
+    } else if (mInputVariableType == Globals::DataLocation::NodeNonHistorical) {
+        KRATOS_ERROR_IF_NOT(rGeometry[0].Has(rIndependentVariable)) << "The Variable " << rIndependentVariable.Name() << " is not available at the nodes of the Geometry to retrieve Table values." << std::endl;
+        for (SizeType i = 0; i < rShapeFunctionVector.size(); ++i) {
+            KRATOS_DEBUG_ERROR_IF_NOT(rGeometry[i].Has(rIndependentVariable)) << "The Variable " << rIndependentVariable.Name() << " is not available at the nodes of the Geometry to retrieve Table values." << std::endl;
+            const double nodal_value = rGeometry[i].GetValue(rIndependentVariable);
+            independent_at_gauss += nodal_value * rShapeFunctionVector[i];
+        }
+    } else if (mInputVariableType == Globals::DataLocation::Element) {
+        KRATOS_ERROR_IF_NOT(rGeometry.Has(rIndependentVariable)) << "The Variable " << rIndependentVariable.Name() << " is not available at the Geometry to retrieve Table values." << std::endl;
+        independent_at_gauss = rGeometry.GetValue(rIndependentVariable);
```

Comment:
I wouldn't do all these variable checks as this function is expected to be accessed each time we access the Gauss pt.

Reply:
should we maintain only the debug one?. This is the only place where we can do this check since we need the geometry

Reply:
Maybe we can do all of them only in debug mode.